turbopack/crates/turbo-tasks-backend/src/backend/operation/update_cell.rs

```diff
@@ -40,18 +40,58 @@ pub enum UpdateCellOperation {
 }
 
 impl UpdateCellOperation {
-    pub fn run(task_id: TaskId, cell: CellId, content: CellContent, mut ctx: impl ExecuteContext) {
+    pub fn run(
+        task_id: TaskId,
+        cell: CellId,
+        content: CellContent,
+        #[cfg(feature = "verify_determinism")] never_equal: bool,
+        #[cfg(not(feature = "verify_determinism"))] _never_equal: bool,
+        mut ctx: impl ExecuteContext,
+    ) {
+        let content = if let CellContent(Some(new_content)) = content {
+            Some(new_content.into_typed(cell.type_id))
+        } else {
+            None
+        };
+
         let mut task = ctx.task(task_id, TaskDataCategory::All);
 
+        let is_stateful = task.has_key(&CachedDataItemKey::Stateful {});
         // We need to detect recomputation, because here the content has not actually changed (even
         // if it's not equal to the old content, as not all values implement Eq). We have to
         // assume that tasks are deterministic and pure.
-        let should_invalidate = ctx.should_track_dependencies()
-            && (task.has_key(&CachedDataItemKey::Dirty {}) ||
+        let assume_unchanged = !ctx.should_track_dependencies()
+            || (!task.has_key(&CachedDataItemKey::Dirty {})
             // This is a hack for the streaming hack. Stateful tasks are never recomputed, so this forces invalidation for them in case of this hack.
-            task.has_key(&CachedDataItemKey::Stateful {}));
+            && !is_stateful);
+
+        let old_content = get!(task, CellData { cell });
+
+        if assume_unchanged {
+            if old_content.is_some() {
+                // Never update cells when recomputing if they already have a value.
+                // It's not expected that content changes during recomputation.
+
+                // Check if this assumption holds.
+                #[cfg(feature = "verify_determinism")]
+                if !is_stateful && !never_equal && content.as_ref() != old_content {
+                    let task_description = ctx.get_task_description(task_id);
+                    let cell_type = turbo_tasks::registry::get_value_type(cell.type_id).global_name;
+                    println!(
```

Comment:
nit: eprintln! though i would actually expect a panic

Reply:
panic is a bit annoying as you want to see all errors of the whole compilation, but a panic would stop the build.


```suggestion
                    eprintln!(
```

Reply:
yeah once more progress is made it might be a good idea to enable this in the `release_with_assertions` build to prevent regressions in ci.  but error printing for now makes sense.